### PR TITLE
[SPARK-20636] Add new optimization rule to transpose adjacent Window expressions.

### PR DESCRIPTION
**Pull request by ptkool**
## What changes were proposed in this pull request?

Add new optimization rule to eliminate unnecessary shuffling by flipping adjacent Window expressions.

## How was this patch tested?

Tested with unit tests, integration tests, and manual tests.

### CODE REVIEW

**Comment by gatorsmile** [comment]
ok to test

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 1]
```diff
@@ -609,6 +610,19 @@ object CollapseWindow extends Rule[LogicalPlan] {
...
 
 /**
+ * Flip Adjacent Window Expressions.
+ * - If the partition spec of the parent Window expression is a subset of the partition spec
+ *   of the child window expression, flip them.
+ */
+object FlipWindow extends Rule[LogicalPlan] {
+  def apply(plan: LogicalPlan): LogicalPlan = plan transformUp {
+    case w1 @ Window(we1, ps1, os1, w2 @ Window(we2, ps2, os2, grandChild))
+        if ps2.containsSlice(ps1) =>
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: This condition might not be enough. `we1` might depend on the outputs of `we2`, right?

Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: You are also changing the order of the columns. You will need to add a projection on top to be sure.

**Comment by hvanhovell** [comment]
retest this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 2]
```diff
@@ -609,6 +610,19 @@ object CollapseWindow extends Rule[LogicalPlan] {
...
 /**
+ * Transpose Adjacent Window Expressions.
+ * - If the partition spec of the parent Window expression is a subset of the partition spec
+ *   of the child window expression, transpose them.
+ */
+object TransposeWindow extends Rule[LogicalPlan] {
+  def apply(plan: LogicalPlan): LogicalPlan = plan transformUp {
+    case w1 @ Window(we1, ps1, os1, w2 @ Window(we2, ps2, os2, grandChild))
+        if ps1.length < ps2.length && ps2.containsSlice(ps1) =>
+      Project(w1.output, Window(we2, ps2, os2, Window(we1, ps1, os1, grandChild)))
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: This probably warrants a follow-up that tries to move projections that are wedged in between two window clauses.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 3]
```diff
@@ -609,6 +610,19 @@ object CollapseWindow extends Rule[LogicalPlan] {
...
 
 /**
+ * Transpose Adjacent Window Expressions.
+ * - If the partition spec of the parent Window expression is a subset of the partition spec
+ *   of the child window expression, transpose them.
+ */
+object TransposeWindow extends Rule[LogicalPlan] {
+  def apply(plan: LogicalPlan): LogicalPlan = plan transformUp {
+    case w1 @ Window(we1, ps1, os1, w2 @ Window(we2, ps2, os2, grandChild))
+        if ps1.length < ps2.length && ps2.containsSlice(ps1) =>
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: You need to check that the windows are independent, e.g.: `w1.references.intersect(w2.windowOutputSet).isEmpty`

Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: We might be able to get a little more milage out of the rule by using `semanticEquals` for comparing the partition expressions, e.g.:
```scala
def sliceSemanticEquals(ps1: Seq[Expression], ps2: Seq[Expression]): Boolean = ps1.zip(ps2).forall {
  case (l, r) => l.semanticEquals(r)
}
...
sliceSemanticEquals(ps1, ps2)
``` 

You could even get more leverage if you do not consider the order of the partition spec.

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameWindowFunctionsSuite.scala [hunk 4]
```diff
@@ -423,4 +423,25 @@ class DataFrameWindowFunctionsSuite extends QueryTest with SharedSQLContext {
       df.select(selectList: _*).where($"value" < 2),
       Seq(Row(3, "1", null, 3.0, 4.0, 3.0), Row(5, "1", false, 4.0, 5.0, 5.0)))
   }
+
+  test("window functions in multiple selects") {
```
Comment (hvanhovell) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameWindowFunctionsSuite.scala]: Why this test? It does not really add anything new.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/TransposeWindowSuite.scala [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+      .window(Seq(sum(c).as('sum_a_1)), partitionSpec1, Seq.empty)
+
+    val analyzed = query.analyze
+    val optimized = Optimize.execute(analyzed)
+
+    val correctAnswer = testRelation
+      .window(Seq(sum(c).as('sum_a_2)), partitionSpec3, Seq.empty)
+      .window(Seq(sum(c).as('sum_a_1)), partitionSpec1, Seq.empty)
+
+    comparePlans(optimized, correctAnswer.analyze)
```
Comment (hvanhovell) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/TransposeWindowSuite.scala]: `comparePlans(optimized, analyzed)`?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/TransposeWindowSuite.scala [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+  val d = testRelation.output(3)
+
+  val partitionSpec1 = Seq(a)
+  val partitionSpec2 = Seq(a, b)
+  val partitionSpec3 = Seq(d)
+
+  val orderSpec1 = Seq(d.asc)
+  val orderSpec2 = Seq(d.desc)
+
+  test("flip two adjacent windows with compatible partitions in multiple selects") {
```
Comment (hvanhovell) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/TransposeWindowSuite.scala]: I am not entirely sure if we need this test.

**Review by hvanhovell — CHANGES_REQUESTED** [review]
@ptkool this looks pretty good. One thing need to be addressed though: we need to factor in that a parent window can depend on its child window.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala (hunk 3 above)
Comment (ptkool) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: Agreed.

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameWindowFunctionsSuite.scala (hunk 4 above)
Comment (ptkool) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameWindowFunctionsSuite.scala]: I'll remove it.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/TransposeWindowSuite.scala (hunk 5 above)
Comment (ptkool) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/TransposeWindowSuite.scala]: Yes.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/TransposeWindowSuite.scala (hunk 6 above)
Comment (ptkool) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/TransposeWindowSuite.scala]: Agreed. I'll remove it.

**Comment by ptkool** [comment]
@hvanhovell @gatorsmile Can you have another look at this?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 7]
```diff
@@ -610,6 +611,25 @@ object CollapseWindow extends Rule[LogicalPlan] {
...
+object TransposeWindow extends Rule[LogicalPlan] {
+  private def compatibleParititions(ps1 : Seq[Expression], ps2: Seq[Expression]): Boolean = {
+    ps1.length < ps2.length && ps2.take(ps1.length).permutations.exists(ps1.zip(_).forall {
+      case (l, r) => l.semanticEquals(r)
+    })
+  }
+
+  def apply(plan: LogicalPlan): LogicalPlan = plan transformUp {
+    case w1 @ Window(we1, ps1, os1, w2 @ Window(we2, ps2, os2, grandChild))
+        if w1.references.intersect(w2.windowOutputSet).isEmpty && compatibleParititions(ps1, ps2) =>
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: No test case covers the condition `w1.references.intersect(w2.windowOutputSet).isEmpty`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 8]
```diff
@@ -610,6 +611,25 @@ object CollapseWindow extends Rule[LogicalPlan] {
...
+ */
+object TransposeWindow extends Rule[LogicalPlan] {
+  private def compatibleParititions(ps1 : Seq[Expression], ps2: Seq[Expression]): Boolean = {
+    ps1.length < ps2.length && ps2.take(ps1.length).permutations.exists(ps1.zip(_).forall {
+      case (l, r) => l.semanticEquals(r)
+    })
+  }
+
+  def apply(plan: LogicalPlan): LogicalPlan = plan transformUp {
+    case w1 @ Window(we1, ps1, os1, w2 @ Window(we2, ps2, os2, grandChild))
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: The expressions in both `we1` and `we2` must be deterministic.

Comment (ptkool) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: Why? This seems overly restrictive to me.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala (hunk 7 above)
Comment (ptkool) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: I will add one.

**Comment by gatorsmile** [comment]
@ptkool Can you address the conflicts? We will review it.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala (hunk 8 above)
Comment (gatorsmile) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: Just to ensure the results are still the same with and without the rule.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 9]
```diff
@@ -726,14 +726,36 @@ object CollapseRepartition extends Rule[LogicalPlan] {
 object CollapseWindow extends Rule[LogicalPlan] {
   def apply(plan: LogicalPlan): LogicalPlan = plan transformUp {
     case w1 @ Window(we1, ps1, os1, w2 @ Window(we2, ps2, os2, grandChild))
-        if ps1 == ps2 && os1 == os2 && w1.references.intersect(w2.windowOutputSet).isEmpty &&
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: I wouldn't include style changes

**Review by gatorsmile — COMMENTED** [review]
LGTM

Thanks! Merged to master.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 10]
```diff
@@ -734,6 +734,28 @@ object CollapseWindow extends Rule[LogicalPlan] {
   }
 }
 
+/**
+ * Transpose Adjacent Window Expressions.
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: why is this rule useful?

**Pull request closed without merging**